tests/failpoints/test_io_error.rs

```diff
@@ -0,0 +1,361 @@
+}
+
+impl ConcurrentWriteContext {
+    fn new(engine: Arc<RaftLogEngine>) -> Self {
+        Self {
+            engine,
+            ths: Vec::new(),
+            ticket: 0,
+            timer: Arc::new(AtomicU64::new(0)),
+        }
+    }
+
+    fn leader_write<F: FnOnce(Result<usize>) + Send + Sync + 'static>(
+        &mut self,
+        mut log_batch: LogBatch,
+        sync: bool,
+        cb: Option<F>,
+    ) {
+        if self.ths.is_empty() {
+            fail::cfg("write_barrier::leader_exit", "pause").unwrap();
+            let engine_clone = self.engine.clone();
+            let ticket = self.ticket;
+            self.ticket += 1;
+            let timer = self.timer.clone();
+            self.ths.push(
+                std::thread::Builder::new()
+                    .spawn(move || {
+                        while ticket != timer.load(std::sync::atomic::Ordering::Acquire) {}
+                        engine_clone.write(&mut LogBatch::default(), false).unwrap();
+                        timer.store(ticket + 1, std::sync::atomic::Ordering::Release);
+                    })
+                    .unwrap(),
+            );
+        }
+        let engine_clone = self.engine.clone();
+        let ticket = self.ticket;
+        self.ticket += 1;
+        let timer = self.timer.clone();
+        self.ths.push(
+            std::thread::Builder::new()
+                .spawn(move || {
+                    while ticket != timer.load(std::sync::atomic::Ordering::Acquire) {}
+                    let r = engine_clone.write(&mut log_batch, sync);
+                    timer.store(ticket + 1, std::sync::atomic::Ordering::Release);
+                    if let Some(f) = cb {
+                        f(r)
+                    }
+                })
+                .unwrap(),
+        );
+    }
+
+    fn follower_write<F: FnOnce(Result<usize>) + Send + Sync + 'static>(
+        &mut self,
+        mut log_batch: LogBatch,
+        sync: bool,
+        cb: Option<F>,
+    ) {
+        assert!(self.ticket >= 2);
+        let engine_clone = self.engine.clone();
+        let ticket = self.ticket;
+        self.ticket += 1;
+        let timer = self.timer.clone();
+        self.ths.push(
+            std::thread::Builder::new()
+                .spawn(move || {
+                    while ticket != timer.load(std::sync::atomic::Ordering::Acquire) {}
+                    let r = engine_clone.write(&mut log_batch, sync);
+                    if let Some(f) = cb {
+                        f(r)
+                    }
+                    timer.store(ticket + 1, std::sync::atomic::Ordering::Release);
+                })
+                .unwrap(),
+        );
+    }
+
+    fn join(&mut self) {
+        fail::remove("write_barrier::leader_exit");
+        for t in self.ths.drain(..) {
+            t.join().unwrap();
+        }
+    }
+}
+
+#[test]
+fn test_open_error() {
+    fail::cfg("log_fd::open::err", "return").unwrap();
+    assert!(catch_unwind_silent(|| {
```

Comment:
This is not a good way to test panic, I can't tell which part is actually panicking from this code.